garage/tf/algos/td3.py

```diff
@@ -0,0 +1,252 @@
+minimum value of two critics instead of one to limit overestimation.
+"""
+
+import numpy as np
+import tensorflow as tf
+import tensorflow.contrib as tc
+
+from garage.misc.overrides import overrides
+from garage.tf.algos import DDPG
+from garage.tf.misc import tensor_utils
+
+
+class TD3(DDPG):
+    """
+    Implementation of TD3.
+
+    Based on https://arxiv.org/pdf/1802.09477.pdf.
+
+    Example:
+        $ python garage/examples/tf/td3_pendulum.py
+    """
+
+    def __init__(self,
+                 env_spec,
+                 qf2,
+                 replay_buffer,
+                 target_update_tau=0.01,
+                 policy_lr=1e-4,
+                 qf_lr=1e-3,
+                 policy_weight_decay=0,
+                 qf_weight_decay=0,
+                 policy_optimizer=tf.train.AdamOptimizer,
+                 qf_optimizer=tf.train.AdamOptimizer,
+                 clip_pos_returns=False,
+                 clip_return=np.inf,
+                 discount=0.99,
+                 max_action=None,
+                 name=None,
+                 **kwargs):
+        """
+        Construct class.
+
+        Args:
+            env(): Environment.
+            qf2(policy): Q function to use
+            target_update_tau(float): Interpolation parameter for doing the
+        soft target update.
+            discount(float): Discount factor for the cumulative return.
+            policy_lr(float): Learning rate for training policy network.
+            qf_lr(float): Learning rate for training q value network.
+            policy_weight_decay(float): L2 weight decay factor for parameters
+        of the policy network.
+            qf_weight_decay(float): L2 weight decay factor for parameters
+        of the q value network.
+            policy_optimizer(): Optimizer for training policy network.
+            qf_optimizer(): Optimizer for training q function network.
+            clip_pos_returns(boolean): Whether or not clip positive returns.
+            clip_return(float): Clip return to be in [-clip_return,
+        clip_return].
+            max_action(float): Maximum action magnitude.
+            name(str): Name of the algorithm shown in computation graph.
+        """
+        self.qf2 = qf2
+
+        super(TD3, self).__init__(
+            env_spec=env_spec,
+            replay_buffer=replay_buffer,
+            target_update_tau=target_update_tau,
+            policy_lr=policy_lr,
+            qf_lr=qf_lr,
+            policy_weight_decay=policy_weight_decay,
+            qf_weight_decay=qf_weight_decay,
+            policy_optimizer=policy_optimizer,
+            qf_optimizer=qf_optimizer,
+            clip_pos_returns=clip_pos_returns,
+            clip_return=clip_return,
+            discount=discount,
+            max_action=max_action,
+            name=name,
+            **kwargs)
+
+    @overrides
+    def init_opt(self):
+        """Init the optimizer."""
```

Comment:
i think this actually builds the loss function (and then later initializes the optimizer)

Reply:
Are you suggesting that the name `init_opt` is not appropriate? I think DDPG also builds the loss function in `init_opt()`

Reply:
it was a remark about the comment, which doesn't seem to say much.